pkg/logql/count_min_sketch.go

```diff
@@ -0,0 +1,265 @@
+package logql
+
+import (
+	"container/heap"
+	"fmt"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+	promql_parser "github.com/prometheus/prometheus/promql/parser"
+
+	"github.com/grafana/loki/v3/pkg/logproto"
+	"github.com/grafana/loki/v3/pkg/logql/sketch"
+	"github.com/grafana/loki/v3/pkg/logql/syntax"
+)
+
+const (
+	CountMinSketchVectorType = "CountMinSketchVector"
+
+	epsilon   = 0.0001
+	delta     = 0.05
+	maxLabels = 10_000
+)
+
+type CountMinSketchVector struct {
+	T int64
+	F *sketch.CountMinSketch
+
+	Metrics []labels.Labels
+}
+
+func (CountMinSketchVector) SampleVector() promql.Vector {
+	return promql.Vector{}
+}
+
+func (CountMinSketchVector) QuantileSketchVec() ProbabilisticQuantileVector {
+	return ProbabilisticQuantileVector{}
+}
+
+func (v CountMinSketchVector) CountMinSketchVec() CountMinSketchVector {
+	return v
+}
+
+func (v *CountMinSketchVector) Merge(right *CountMinSketchVector) (*CountMinSketchVector, error) {
+	err := v.F.Merge(right.F)
+	if err != nil {
+		return v, err
+	}
+
+	// TODO: merge labels
```

Comment:
I implemented merging of the heaps in the existing sketch/topk that we could use as a basis. It might not be very fast, it essentially replays all things that are in the heap from each sharded sketch into a new heap.